client/cli/src/lib.rs

```diff
@@ -210,6 +210,27 @@ pub trait SubstrateCli: Sized {
 		command.init::<Self>()?;
 		Runner::new(self, command)
 	}
+
+	/// Default RPC HTTP port.
+	///
+	/// By default this is 9933.
+	fn default_rpc_http_port() -> u16 {
```

Comment:
I don't really understand why he have these ports in this trait here?

Reply:
I explained in detail here: https://github.com/paritytech/substrate/pull/6250#discussion_r435292134

TL;DR: This is to make sure the default value is the same everywhere (on all subcommands). SubstrateCli has kind of "static"/"constant"-like stuff while ConfigurationCli know how to transform things for a subcommand.

Reply:
I'm not entirely sold on this solution. This distinction between these traits seems rather random. And what is with other default values? Why are they not override-able?

Reply:
Me neither :sweat_smile: initially when I did the refactoring I wanted to have only one trait but it wasn't possible.

Some things work using the parameters provided by the user (CliConfiguration) like: getting the database parameters.

Some other things work using kinda global values (SubstrateCli) like: impl name, impl version.

Some are a bit of both like: getting the chain spec value from user input but process the input the same way for the entire Cli.

In CliConfiguration there are no function that return what should be the default value of something. They are all kinda getter to the user input. Here what we override is the default port of something.

Well, either way I'm out of argument so I will let you decide what is best among those options and I will adapt the code accordingly:

1. Move the default port methods to CliConfiguration and use Result and &self (what I did originally)
2. Move the default port methods to CliConfiguration and not use Result but use &self (what @expenses proposed)
3. Move the default port methods to SubstrateCli and not use Result nor &self (what is it right now)

(Note: there is no option for using CliConfiguration without Result and without &self because &self is required to dispatch to the right subcommand if used in CliConfiguration. This is not the case for SubstrateCli because the trait is kinda global to the entire Cli app.)

cc @seunlanlege (because related to encapsulation OCD)

> The alternative is to have these methods on CliConfiguration and use proc macro magic to override functions

I have been asked to avoid proc macros in the API as much as possible, so this won't even be an option for me.